### PR TITLE
Engine signer password

### PR DESCRIPTION
**Pull request by joojis**

### CODE REVIEW

File: codechain/main.rs [hunk 1]
```diff
@@ -198,22 +198,30 @@ fn new_miner(config: &config::Config, spec: &Spec, ap: Arc<AccountProvider>) ->
...
+    let miner = Miner::new(miner_options, spec, Some(ap.clone()));
 
-    let miner = Miner::new(miner_options, spec, Some(ap));
-
-    let author = config.mining.author.unwrap_or(address);
     if miner.can_produce_work_package() {
-        miner.set_author(author);
+        let author = config.mining.author;
+        match author {
+            Some(author) => {
```
Comment (kseo) [inline codechain/main.rs]: Remove redundant curly braces.

File: codechain/main.rs [hunk 2]
```diff
@@ -198,22 +198,30 @@ fn new_miner(config: &config::Config, spec: &Spec, ap: Arc<AccountProvider>) ->
...
+                miner.set_author(author);
+            }
+            None => return Err("mining.author is not defined".to_string()),
```
Comment (kseo) [inline codechain/main.rs]: defined -> specified

File: codechain/main.rs [hunk 3]
```diff
@@ -198,22 +198,30 @@ fn new_miner(config: &config::Config, spec: &Spec, ap: Arc<AccountProvider>) ->
...
+        }
     } else {
+        if config.mining.engine_signer.is_none() {
```
Comment (kseo) [inline codechain/main.rs]: Use match.

File: codechain/main.rs [hunk 4]
```diff
@@ -198,22 +198,30 @@ fn new_miner(config: &config::Config, spec: &Spec, ap: Arc<AccountProvider>) ->
...
+            return Err("mining.engine_signer is not defined".to_string())
```
Comment (kseo) [inline codechain/main.rs]: defined -> specified

File: codechain/main.rs [hunk 5]
```diff
@@ -221,8 +222,14 @@ fn new_miner(config: &config::Config, spec: &Spec, ap: Arc<AccountProvider>) ->
             }
             Err(e) => return Err(format!("Error while checking whether engine_signer is in AccountProvider: {:?}", e)),
         }
-        // FIXME: Don't hardcode password.
-        miner.set_engine_signer(engine_signer, "password".to_string()).map_err(|err| format!("{:?}", err))?;
+        if config.mining.password_path.is_none() {
+            return Err("mining.password_path is not defined".to_string())
```
Comment (kseo) [inline codechain/main.rs]: defined -> specified

File: codechain/main.rs [hunk 6]
```diff
@@ -221,8 +222,14 @@ fn new_miner(config: &config::Config, spec: &Spec, ap: Arc<AccountProvider>) ->
...
+        }
+        if let Ok(password) = fs::read_to_string(config.mining.password_path.clone().unwrap()) {
```
Comment (kseo) [inline codechain/main.rs]: What if the file does not exist?

File: codechain/main.rs [hunk 7]
```diff
@@ -198,22 +198,30 @@ fn new_miner(config: &config::Config, spec: &Spec, ap: Arc<AccountProvider>) ->
...
+            }
+            None => return Err("mining.author is not defined".to_string()),
+        }
     } else {
+        if config.mining.engine_signer.is_none() {
+            return Err("mining.engine_signer is not defined".to_string())
+        }
+        let engine_signer = config.mining.engine_signer.unwrap();
+        match ap.has_account(engine_signer) {
+            Ok(has_account) => {
```
Comment (kseo) [inline codechain/main.rs]: Use guards.
https://doc.rust-lang.org/1.5.0/book/patterns.html#guards

File: codechain/main.rs [hunk 8]
```diff
@@ -208,22 +209,26 @@ fn new_miner(config: &config::Config, spec: &Spec, ap: Arc<AccountProvider>) ->
...
+                            Ok(password) => {
+                                miner.set_engine_signer(engine_signer, password).map_err(|err| format!("{:?}", err))?
+                            }
+                            Err(_) => return Err(format!("Failed to read the password file")),
+                        },
                     }
+                    miner.set_engine_signer(engine_signer, "password".to_string()).map_err(|err| format!("{:?}", err))?;
                 }
-            }
+                Err(e) => {
```
Comment (kseo) [inline codechain/main.rs]: Remove braces.

Comment (joojis) [reply codechain/main.rs]: `cargo fmt` says that the braces are needed.

File: codechain/main.rs [hunk 9]
```diff
@@ -209,22 +210,26 @@ fn new_miner(config: &config::Config, spec: &Spec, ap: Arc<AccountProvider>) ->
...
-                        return Err(format!("Error while checking whether engine_signer is in AccountProvider: {:?}", e))
+            Some(engine_signer) => match ap.has_account(engine_signer) {
+                Ok(has_account) if !has_account => {
+                    return Err("mining.engine_signer is not found in AccountProvider".to_string())
+                }
+                Ok(..) => {
+                    match config.mining.password_path {
+                        None => return Err("mining.password_path is not specified".to_string()),
+                        Some(ref password_path) => match fs::read_to_string(password_path) {
+                            Ok(password) => {
```
Comment (kseo) [inline codechain/main.rs]: Remove braces.

**Pull request merged**